trapz_errors.py

```diff
@@ -0,0 +1,518 @@
+    integral = calculate_integral(points)
+    total_error = calculate_total_error(points, conservative=conservative)
+    print(f"Integral: {integral:.{precision}f} ± {total_error:.{precision}f}")
+
+
+def print_errors(points: np.ndarray, precision: int = 3, conservative: bool = False):
+    y_uncertainty = calculate_y_value_uncertainty(points)
+    print(f"Error from y-value uncertainty: ± {y_uncertainty:.{precision}f}")
+    integral = calculate_integral(points)
+    interval_errors, _ = calculate_interval_errors(points)
+    total_truncation_error = np.abs(np.sum(interval_errors))
+    print(
+        f"Estimated total truncation error: {total_truncation_error:.{precision}f}")
+    total_error = calculate_total_error(points, conservative=conservative)
+    print(f"Integral: {integral:.{precision}f} ± {total_error:.{precision}f}")
+
+
+def get_largest_error_points(
+    points_to_assess: np.ndarray,
+    residual_error: float,
+    largest_absolute_error: float,
+    convergence_rate_scale_factor: float = 1,
+    conservative: bool = False,
+) -> np.ndarray:
+    points = sorted(points_to_assess, reverse=True, key=lambda x: x[-1])
+    largest_error_points = []
+
+    scale_factor = 0.75 / convergence_rate_scale_factor
+    while residual_error >= 0 and points:
+        point = points.pop(0)
+        largest_error_points.append(point)
+        abspoint = abs(point)
+        if conservative and np.allclose(point[-1], largest_absolute_error):
+            residual_error -= abspoint[-1]
+        else:
+            # COMMENT COPIED FROM Martin Stroet, February 5th, 2018 1:36pm ||  92df53f
+            # Since addition of a point to an interval reduces error by a factor of 1/4 according
+            # to the truncation error estimate method we're using => (e - e/4) = 0.75*e.
+            # We simply assume the same is true for reducing the uncertainty on an existing point.
+            # The convergence rate scaling factor allows for additional control over the number of points
+            # that will be added on each iteration.
+            residual_error -= scale_factor * abspoint[-1]
+
+    return np.array(largest_error_points)
+
+
+def get_points_to_add_and_extend(
+    points: np.ndarray,
+    residual_error: float,
+    conservative: bool = False,
+    convergence_rate_scale_factor: float = 1,
+) -> Tuple[np.ndarray, np.ndarray]:
+    interval_points = calculate_interval_points(points)
+    interval_errors, _ = calculate_interval_errors(points)
+    interval_points = np.concatenate([interval_points.T, [interval_errors]]).T
+
+    points_with_error = np.array(points)
+    points_with_error[:, 2] = calculate_point_errors(
+        points[:, 0], points[:, 2])
+    points_to_assess = np.concatenate([points_with_error, interval_points])
+    # largest_absolute_error = np.max(np.abs(interval_points[:, 2]))
+    largest_absolute_error = np.abs(interval_points[0])[2]
```

Comment:
These lines correspond to:

https://github.com/ATB-UQ/trapz_errors/blob/05019d9dd6a9e3af4d2e406c8069564bb4c469fc/reduce_error.py#L13-L15

While the original script mentions that `# since pts are already sorted we can simply filter out all non-gap points and take the first which remains`, I am not sure that it is true that the points are already sorted. As far as I can tell, they are ordered by x (second column), but not error (first column). If I print out what `reduce_error.py` selects, it is the first point: `(-0.05120933333333332, 0.0125, 'gap')`. All the comments in `reduce_error.py` seem to treat `largest_gap_error_pt` as a special case of the largest error, and it does change the output, so I just wanted to check this part as well!

Reply:
This is either a bug or at the very least extremely flimsy... I pushed a version with this corrected.